omhttp/handler.go

```diff
@@ -0,0 +1,150 @@
+package omhttp
+
+import (
+	"compress/gzip"
+	"net/http"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/bsm/openmetrics"
+)
+
+// DefaultHandler is a short-cut for NewHandler(openmetrics.DefaultRegistry(), opts...).
+func DefaultHandler(opts ...HandlerOption) http.Handler {
+	return NewHandler(nil, opts...)
+}
+
+type handler struct {
+	reg *openmetrics.Registry
+}
+
+// NewHandler inits a new handler.
+func NewHandler(reg *openmetrics.Registry, opts ...HandlerOption) http.Handler {
+	var c handlerConfig
+	for _, o := range opts {
+		o.update(&c)
+	}
+
+	if reg == nil {
+		reg = openmetrics.DefaultRegistry()
+	}
+
+	var h http.Handler = handler{reg: reg}
+	if skip := c.noCompression; !skip {
+		h = withCompression(h)
+	}
+	if n := c.limitConcurrency; n > 0 {
+		h = limitConcurrency(h, n)
+	}
+	return h
+}
+
+func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	header := w.Header()
+	header.Set("Content-Type", openmetrics.ContentType)
+
+	n, err := h.reg.WriteTo(w)
+	if err != nil && n == 0 {
+		msg := "An internal error has occurred:\n\n" + err.Error()
+		http.Error(w, msg, http.StatusInternalServerError)
+	}
+}
+
+// ----------------------------------------------------------------------------
+
+type handlerConfig struct {
+	noCompression    bool
+	limitConcurrency int
+}
+
+// A HandlerOption configures the handler.
+type HandlerOption interface {
+	update(*handlerConfig)
+}
+
+type inlineHandlerOption func(*handlerConfig)
+
+func (f inlineHandlerOption) update(c *handlerConfig) { f(c) }
+
+// NoCompression disables default compression of the response
+// body based on Accept-Encoding request headers.
+func NoCompression() HandlerOption {
+	return inlineHandlerOption(func(c *handlerConfig) { c.noCompression = true })
+}
+
+// LimitConcurrency limits the number of concurrent requests that
+// can be made to the handler. The endpoint will start responding
+// with 503 Service Unavailable once this limit is exceeded.
+func LimitConcurrency(n int) HandlerOption {
+	return inlineHandlerOption(func(c *handlerConfig) { c.limitConcurrency = n })
+}
```

Comment:
Minor, but why it's needed at all (use-case)?

Reply:
You normally want some throttling. You don't want to (accidentally) receive too many concurrent requests on this endpoint as every requests starts a (blocking) collection cycle.

Reply:
inspired by https://pkg.go.dev/github.com/prometheus/client_golang@v1.10.0/prometheus/promhttp#HandlerOpts